03-GettingStarted/06-http-streaming/README.md

```diff
@@ -0,0 +1,427 @@
+
+---
+
+## 3. Streaming in MCP
+
+In MCP, streaming is not about sending the main response in chunks, but about sending **notifications** to the client while a tool is processing a request. These notifications can include progress updates, logs, or other events.
+
+#### Key differences from traditional streaming
+- The main result is still sent as a single response
+- Notifications are sent as separate messages during processing
+- The client must be able to handle and display these notifications
+
+#### What is a Notification?
+A notification is a message sent from the server to the client to inform about progress, status, or other events during a long-running operation. Notifications improve transparency and user experience.
+
+##### Why use notifications?
+- Keep users informed about progress
+- Enable real-time feedback and better UX
+- Useful for debugging and monitoring
+
+---
+
+## 4. Implementing Notifications in MCP
+
+### Server-side: Sending Notifications
+To send notifications from your MCP tool, use the context object (usually `ctx`) to call methods like `ctx.info()` or `ctx.log()`. These send messages to the client as the server processes a request.
+
+```python
+@mcp.tool(description="A tool that sends progress notifications")
+async def process_files(message: str, ctx: Context) -> TextContent:
+    await ctx.info("Processing file 1/3...")
+    await ctx.info("Processing file 2/3...")
+    await ctx.info("Processing file 3/3...")
+    return TextContent(type="text", text=f"Done: {message}")
+```
+
+### Client-side: Receiving Notifications
+The client must implement a message handler to process and display notifications as they arrive.
+
+```python
+async def message_handler(message):
+    if isinstance(message, types.ServerNotification):
+        print("NOTIFICATION:", message)
+    else:
+        print("SERVER MESSAGE:", message)
+```
+
+To enable notifications, ensure your server uses a streaming transport (like `streamable-http`) and your client implements a message handler to process notifications.
+
+---
+
+## 5. Streamable HTTP Transport
+
+### What is Streamable HTTP?
+Streamable HTTP is a transport mechanism in MCP that uses HTTP POST requests and supports streaming notifications (such as progress updates) from the server to the client. It is designed for modern web and cloud environments.
+
+### How It Works
+- The client sends a request to the server using HTTP.
+- The server can send notifications (e.g., progress, logs) back to the client while processing the request.
+- The final response is sent when processing is complete.
+
+### Benefits Over SSE and stdio
+- **Better scalability**: Works well with load balancers and cloud deployments.
+- **Stateless and stateful modes**: Supports both, with resumability.
+- **Rich notifications**: Send progress, logs, and other events during processing.
+- **Standard HTTP**: Easier to integrate with existing infrastructure.
+
+### Implementation Details
```

Comment:
add lead in sentences, try to balance this out with more paragraphs, a bullet list should enhance the message of a header but not replace normal text